### ISSUE TITLE
Newly added namespace is not listed

### ISSUE THREAD

**Issue by hhovsepy** [kiali/kiali#2482]
Happens on master.
Add a new bookinfo and patch kiali to have it in accessible namespaces.
It is replicated into kiali configmap, but still not displayed in Kiali UI.

**Comment by jmazzitelli** [comment]
Changing the configmap under the covers REQUIRES the Kiali pod to be restarted. so if you merely changed the configmap, that is not enough.

However, if you are saying the CR was changed, that should do what you want because the operator should be restarting the Kiali pod.

Can you provide replication steps? Are you modifying the CR?

Can you provide Kiali Operator logs?

**Comment by hhovsepy** [comment]
[kiali-operator-896bc595c-dsj9m-ansible.log](https://github.com/kiali/kiali/files/4287803/kiali-operator-896bc595c-dsj9m-ansible.log)
[kiali-test-mesh-operator-9cf59bddf-rngcw-ansible.log](https://github.com/kiali/kiali/files/4287804/kiali-test-mesh-operator-9cf59bddf-rngcw-ansible.log)

**Comment by hhovsepy** [comment]
It seems the issue is in cache, when it is disabled, the bookinfo namespace is recognized.

**Comment by lucasponce** [comment]
> It seems the issue is in cache, when it is disabled, the bookinfo namespace is recognized.

What do you have in cache_namespaces ? in Kiali CR ?

**Comment by jmazzitelli** [comment]
Just for completeness, can you post the "accessible_namespaces" section of the Kiali CR as well as the section that @lucasponce just asked for - "cache_namespaces"?

**Comment by hhovsepy** [comment]
[configmap-kiali.txt](https://github.com/kiali/kiali/files/4287980/configmap-kiali.txt)
[kiali-kiali.txt](https://github.com/kiali/kiali/files/4287981/kiali-kiali.txt)

**Comment by jmazzitelli** [comment]
> [configmap-kiali.txt](https://github.com/kiali/kiali/files/4287980/configmap-kiali.txt)
> [kiali-kiali.txt](https://github.com/kiali/kiali/files/4287981/kiali-kiali.txt)

OK, those look normal - so I'll assume the problem is the cache unless I'm told otherwise.

**Comment by lucasponce** [comment]
> OK, those look normal - so I'll assume the problem is the cache unless I'm told otherwise.

Probably, I will take a look.

**Comment by jmazzitelli** [comment]
> kiali-operator-896bc595c-dsj9m-ansible.log

I don't see in this log that accessible_namespaces is set to "bookinfo2". Are you sure these are the correct logs? Because the configmap you posted DOES have it, but the operator logs doesn't show it added it.

**Comment by hhovsepy** [comment]
Have build pre-cache version and the problem exists partly:
bookinfo is recognized, but bookinfo2 is not, even though it is set in accessible namespaces of kiali configmap via operator and restarted.

**Comment by lucasponce** [comment]
@hhovsepy I think bookinfo2 has not the annotation:

```
label_selector: kiali.io/member-of=istio-system
```

then Kiali code will skip it:

```
projects, err2 := in.k8s.GetProjects(labelSelector)
```

**Comment by hhovsepy** [comment]
This is on fresh setup environment, both namespaces have:
istio-injection=enabled
kiali-test-mesh-operator=owned
kiali.io/member-of=istio-system

**Comment by jmazzitelli** [comment]
@hhovsepy perhaps it is best to publish your exactly replication procedures. e.g. What does the Kiali CR look like when you first create it - then what commands do you do when modifying the Kiali CR? What do these Kiali CRs look like? Are you changing Kiali CRs quickly or do you wait for the deployments to finish before changing the Kiali CRs?

Finally (and this might be the important one) ... does the bookinfo2 namespace ACTUALLY EXIST at the time you add it to the Kiali CR and when the operator processes it? Or do you add bookinfo2 to  Kiali CR AND THEN create bookinfo2 namespace? If you create it AFTER adding it to the  Kiali CR, then that is a problem - there would be no namespace to add the label.

**Comment by hhovsepy** [comment]
1. Setup Istio and Kiali on fresh installed environment.
Kiali CR does NOT have bookinfo and bookinfo2 in accessible namespaces.
Wait until Kiali UI is up and accessible.
2. Install bookinfo.
[bookinfo_install.txt](https://github.com/kiali/kiali/files/4291869/bookinfo_install.txt)
Kiali CR got bookinfo in accessible namespaces, configmap is updated, kiali pod is restarted.
Wait until Kiali UI displays bookinfo.
3. Install bookinfo2.
[bookinfo2_install.txt](https://github.com/kiali/kiali/files/4291882/bookinfo2_install.txt)
Kiali CR got bookinfo and bookinfo2 in accessible namespaces, configmap is updated, kiali pod is restarted.
Here Kiali UI does not show bookinfo2.

**Comment by hhovsepy** [comment]
4. Manually restart the kiali pod. bookinfo2 appears in UI.

**Comment by hhovsepy** [comment]
5. Install bookinfo3
[bookinfo3_install.txt](https://github.com/kiali/kiali/files/4292131/bookinfo3_install.txt)
"kiali.io/member-of=istio-system" label IS NOT set.
![Screenshot from 2020-03-05 10-51-16](https://user-images.githubusercontent.com/604313/75969448-4362a180-5ecf-11ea-981f-0dbaf4cd570e.png)

Wait for 15 mins.
Kiali CR is updated (bookinfo3 is there): 
[kiali-kiali.txt](https://github.com/kiali/kiali/files/4292138/kiali-kiali.txt)
Kiali-operator log: 
[kiali-operator-69f5849f77-pxk2k-ansible(1).log](https://github.com/kiali/kiali/files/4292145/kiali-operator-69f5849f77-pxk2k-ansible.1.log)
Kiali configmap is not updated (bookinfo3 is NOT there): 
[configmap-kiali.txt](https://github.com/kiali/kiali/files/4292154/configmap-kiali.txt)
Kiali pod is not restarted. Bookinfo3 is not listed in kiali UI.

**Comment by jmazzitelli** [comment]
> Kiali-operator log:
> [kiali-operator-69f5849f77-pxk2k-ansible(1).log](https://github.com/kiali/kiali/files/4292145/kiali-operator-69f5849f77-pxk2k-ansible.1.log)

That doesn't even have "bookinfo2" in it. So something is weird here that bookinfo2 and bookinfo3 namespaces aren't even seen by the operator when doing step 5.

**Comment by jmazzitelli** [comment]
Take a look at the last operator logs posted (for step 5):

```
TASK [default/kiali-deploy : Find all namespaces (this is limited to what the operator has permission to see)] ***
task path: /opt/ansible/roles/default/kiali-deploy/tasks/main.yml:375
/usr/local/lib/python3.6/site-packages/kubernetes/config/kube_config.py:509: YAMLLoadWarning: calling yaml.load() without Loader=... is deprecated, as the default Loader is unsafe. Please read https://msg.pyyaml.org/load for full details.
  config_dict=yaml.load(f),
ok: [localhost] => {"ansible_facts": {"all_namespaces": ["bookinfo", "default", "istio-system", "kiali-operator", "kiali-test-mesh-operator", "kube-node-lease", "kube-public", "kube-system", "openshift", "openshift-apiserver", "openshift-apiserver-operator", "openshift-authentication", "openshift-authentication-operator", "openshift-cloud-credential-operator", "openshift-cluster-machine-approver", "openshift-cluster-node-tuning-operator", "openshift-cluster-samples-operator", "openshift-cluster-storage-operator", "openshift-cluster-version", "openshift-config", "openshift-config-managed", "openshift-console", "openshift-console-operator", "openshift-controller-manager", "openshift-controller-manager-operator", "openshift-dns", "openshift-dns-operator", "openshift-etcd", "openshift-image-registry", "openshift-infra", "openshift-ingress", "openshift-ingress-operator", "openshift-insights", "openshift-kni-infra", "openshift-kube-apiserver", "openshift-kube-apiserver-operator", "openshift-kube-controller-manager", "openshift-kube-controller-manager-operator", "openshift-kube-scheduler", "openshift-kube-scheduler-operator", "openshift-machine-api", "openshift-machine-config-operator", "openshift-marketplace", "openshift-monitoring", "openshift-multus", "openshift-network-operator", "openshift-node", "openshift-openstack-infra", "openshift-operator-lifecycle-manager", "openshift-operators", "openshift-ovirt-infra", "openshift-sdn", "openshift-service-ca", "openshift-service-ca-operator", "openshift-service-catalog-apiserver-operator", "openshift-service-catalog-controller-manager-operator", "openshift-user-workload-monitoring"]}, "changed": false}
```

There *is no* bookinfo3 OR bookinfo2.

So for some reason, either your bookinfo2 and bookinfo3 were deleted by the time the operator ran this OR the operator was given a role that was not given access to see those namespaces (which should not be if the operator was installed via OLM because it should have cluster-admin rights to see all namespaces).

But whatever the reason, the operator CANNOT see bookinfo2 and bookinfo3 during step 5 and that is why in step 5 `Kiali configmap is not updated (bookinfo3 is NOT there):` and `Bookinfo3 is not listed in kiali UI`

**Comment by hhovsepy** [comment]
The process of installing bookinfos was the same for all 3 cases. And the user has cluster admin rights.
But will retry to remove bookinfos and reinstall them again.

**Comment by hhovsepy** [comment]
Manually deleted 3 bookinfos from accessible_namespaces of Kiali CR.
Kiali configmap is not updated, pod is not restarted.
Kiali CR:
[kiali-kiali.txt](https://github.com/kiali/kiali/files/4293095/kiali-kiali.txt)
Kiali configmap:
[configmap-kiali.txt](https://github.com/kiali/kiali/files/4293105/configmap-kiali.txt)
kiali-operator logs are not changed:
[kiali-operator-69f5849f77-pxk2k-ansible.log](https://github.com/kiali/kiali/files/4293115/kiali-operator-69f5849f77-pxk2k-ansible.log)

**Comment by jmazzitelli** [comment]
are you sure you are editing the correct CR?
It doesn't make any sense that editing the CR doesn't change anything.
And the logs again show only "bookinfo" and "default" as the only namespaces in the cluster (other than all the openshift-* ones).

**Comment by jmazzitelli** [comment]
I also see a "kiali-operator" namespace here - are you testing the operator being installed using the "getLatestKialiOperator" script? I thought you were testing with the operator installed via OLM. How did you install the operator?

**Comment by hhovsepy** [comment]
edited CR is the correct one, kiali-operator is installed via the "deploy-kiali-operator.sh" script.
The steps of installing is exactly the same as described in another issue we see (1-3 steps):
https://github.com/kiali/kiali/issues/2484
But here it seems that problem is not constantly happening. Another bookinfo4 installation went fine.
Can it be because of cluster (OCP 4.3, 3 masters + 2 nodes) poor performance?

**Comment by jmazzitelli** [comment]
I just tested with a master branch build. So the cache is enabled (I confirmed this by looking at the configmap).

I again tested the same was as before:

1. Install bookinfo in namespace "bookinfo"
2. Edit Kiali CR and add "bookinfo" to accessible_namespaces
3. Watch for the label to be created in the bookinfo namespace
4. Watch for the Kiali pod to restart
5. See the Kiali UI display the bookinfo namespace in the Overview page (takes a while for that to happen, but no more than a minute or so).

I repeat steps 1-5 with "bookinfo2" (it worked) and then with "bookinfo3" (it worked, too)

All seems to work.

**Comment by hhovsepy** [comment]
I tried to retest but was not able to reach into the point of this issue, because bookinfo2 installation problems.
1. Installing bookinfo works fine and Kiali recognizes that. CR is updated, kiali configmap is updated, bookinfo is in Kiali UI.
2. Install bookinfo2. And here "kiali-test-mesh-operator" does not see bookinfo2 CR creation, even after several minutes. And only after "kiali-test-mesh-operator" pod gets manually recreated, the bookinfo2 is installed and appears on Kiali UI.
So the initial problem described in this issue is similar to the problem I see with bookinfo's 'kiali-test-mesh-operator': When the CR is edited or created, the operator does not recognize that (at least in ~ 10mins).

**Comment by lucasponce** [comment]
@hhovsepy @jmazzitelli let me know if this issue is still valid or if we can't reproduce it correctly and we want to put it on hold and re-open when we have a way to reproduce it.

**Comment by hhovsepy** [comment]
Closing this issue as it did not happen to me so far.

**Comment by RachnaDodia** [comment]
You can delete service mesh member roll created via operator. Recreate it and add project bookinfo as member
